dbms/src/Storages/DeltaMerge/File/DMFileReader.cpp

```diff
@@ -703,4 +708,59 @@ void DMFileReader::addSkippedRows(UInt64 rows)
         break;
     }
 }
+
+void DMFileReader::initAllMatchBlockInfo()
+{
+    const auto & pack_res = pack_filter.getPackResConst();
+    const auto & pack_stats = dmfile->getPackStats();
+
+    // Get continuous packs with RSResult::All
+    auto get_all_match_block = [&](size_t start_pack) {
+        size_t count = 0;
+        size_t rows = 0;
+        for (size_t i = start_pack; i < pack_res.size(); ++i)
+        {
+            if (!pack_res[i].allMatch() || rows >= rows_threshold_per_read)
+                break;
+
+            ++count;
+            rows += pack_stats[i].rows;
+        }
+        return std::make_pair(count, rows);
+    };
+
+    for (size_t i = 0; i < pack_res.size();)
+    {
+        if (!pack_res[i].allMatch())
+        {
+            ++i;
+            continue;
+        }
+        auto [count, rows] = get_all_match_block(i);
+        // Do not read block too small, it may hurts performance
+        if (rows >= rows_threshold_per_read / 2)
+            all_match_block_infos.emplace(i, count);
+        i += count;
```

Comment:
```suggestion
        auto [pack_count, rows] = get_all_match_block(i);
        // Do not read block too small, it may hurts performance
        if (rows >= rows_threshold_per_read / 2)
            all_match_block_infos.emplace(i, pack_count);
        i += pack_count;
```